File: testbench/virtual_tr/lm32_code/bin2mif.c
#include <stdio.h>
#include <stdint.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

int main() {
	int fd = open("firmware.bin", O_RDONLY);
	printf("DEPTH = 32768;\n"
			"WIDTH = 32;\n"
			"ADDRESS_RADIX = HEX;\n"
			"DATA_RADIX = HEX;\n"
			"CONTENT\n"
			"BEGIN\n");

	for (int i = 0; i < 32768; ++i) {
		uint32_t x, word;
		int result = read(fd, &x, 4);
		if (result != 4) {
			word = 0;
		} else {
			word  = (x>>24) & 0x000000ff;
			word |= (x<<24) & 0xff000000;
			word |= (x>>8)  & 0x0000ff00;
			word |= (x<<8)  & 0x00ff0000;
		}
		printf("%x : %08lx;\n", i, word);
	}
	printf("END;\n");


}

File: modules/neorv32/src/sw/wishbone_test/bin2mif.py
import sys

# Parameter
BIN_FILE = "program.bin"
MIF_FILE = "program.mif"
WORD_WIDTH = 32  # bits
ADDRESS_RADIX = "HEX"
DATA_RADIX = "HEX"

# Read BIN file
with open(BIN_FILE, "rb") as f:
    binary = f.read()

# Slice into 4 byte words
words = []
for i in range(0, len(binary), 4):
    chunk = binary[i:i+4]
    while len(chunk) < 4:
        chunk += b'\x00'  # padding
    word = int.from_bytes(chunk, byteorder='little')
    words.append(word)

depth = len(words)

# Write MIF file
with open(MIF_FILE, "w") as f:
    f.write(f"WIDTH={WORD_WIDTH};\n")
    f.write(f"DEPTH={depth};\n\n")
    f.write(f"ADDRESS_RADIX={ADDRESS_RADIX};\n")
    f.write(f"DATA_RADIX={DATA_RADIX};\n\n")
    f.write("CONTENT BEGIN\n")

    for addr, word in enumerate(words):
        f.write(f"  {addr:02X} : {word:08X};\n")

    f.write("END;\n")

print(f"Done: {MIF_FILE} (Words: {depth})")
